internals/idkey/idkey.go

```diff
@@ -0,0 +1,265 @@
+// the leaf element of the key driectory if it does not yet exist.
+func (k *IDKey) createDir() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return expectPermission(k.keyDir, 0o700)
+	}
+
+	// Create the leaf directory node with 0o700 permissions.
+	return os.Mkdir(k.keyDir, 0o700)
+}
+
+// Public implements the crypto.Signer.
+func (k *IDKey) Public() crypto.PublicKey {
+	signer := k.key.(crypto.Signer)
+	return signer.Public()
+}
+
+// Sign implements the crypto.Signer.
+func (k *IDKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	signer := k.key.(crypto.Signer)
+	return signer.Sign(rand, digest, opts)
+}
+
+// newEd25519 generates a new ed25519 key.
+func (k *IDKey) newEd25519() error {
+	var err error
+	_, k.key, err = ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// load loads the private identity key from storage.
+func (k *IDKey) load() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil || !exists {
+		return fmt.Errorf("directory %q is not accessible", k.keyDir)
+	}
+	err = expectPermission(k.keyDir, 0o700)
+	if err != nil {
+		return err
+	}
+	pemPath := filepath.Join(k.keyDir, identityKeyFile)
+	err = expectPermission(pemPath, 0o600)
+	if err != nil {
+		return err
+	}
+	// Load the key.
+	pemData, err := os.ReadFile(pemPath)
+	if err != nil {
+		return err
+	}
+	for {
+		var block *pem.Block
+		block, pemData = pem.Decode(pemData)
+		if block == nil {
+			break
+		}
+		switch block.Type {
+		case "PRIVATE KEY":
+			k.key, err = x509.ParsePKCS8PrivateKey(block.Bytes)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown PEM block %q in file %q", block.Type, pemPath)
+		}
+	}
+	if k.key == nil {
+		return fmt.Errorf("empty PEM file %q", pemPath)
+	}
+	return nil
+}
+
+// save saves the private identity key to storage.
+func (k *IDKey) save() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil {
+		return fmt.Errorf("directory %q is not accessible", k.keyDir)
+	}
+	if exists {
+		err = expectPermission(k.keyDir, 0o700)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Create the leaf directory node with 0700 permissions.
+		err = os.Mkdir(k.keyDir, 0o700)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create the new private identity file.
+	pemPath := filepath.Join(k.keyDir, identityKeyFile)
+	pemFile, err := os.OpenFile(pemPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
+	}()
+	keyBytes, err := x509.MarshalPKCS8PrivateKey(k.key)
+	if err != nil {
+		return err
+	}
+	pemPrivateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: keyBytes,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemPrivateBlock); err != nil {
```

Comment:
I guess this is intentionally creating a new `err` variable? In which case, it's kinda subtle -- please comment why.

Reply:
My bad. In fact the defer was broken, it was trying to update the _named_ error return, that was not named. I added a named error return, and added a comment inside defer.

I simplified it a little, only doing the close in the defer, which makes more sense. Let me know what you think.

Reply:
Resolving.